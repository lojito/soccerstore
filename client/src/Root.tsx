import { type ReactNode } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";
import reducers from "./reducers";

type Props = {
  children: ReactNode;
  initialStoreValue?: {
    favorites: number[];
    cart: number[];
  };
};

// eslint-disable-next-line react-refresh/only-export-components
export default ({ children, initialStoreValue }: Props) => {
  const store = createStore(
    reducers,
    initialStoreValue as unknown as Partial<{ favorites: never; cart: never }>
  );

  return (
    <BrowserRouter>
      <Provider store={store}>{children}</Provider>
    </BrowserRouter>
  );
};
