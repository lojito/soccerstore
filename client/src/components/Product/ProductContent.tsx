import { type FC, type ReactNode } from "react";
import { styled } from "styled-components";

const DivWrapper = styled.div`
  color: #2d2f31;
  font-weight: bold;
  font-size: 1.6rem;
  margin-bottom: 5px;
  text-align: left;
  width: 225px;
  margin: 0 auto;
`;

type Props = {
  children: ReactNode;
};

const ProductContent: FC<Props> = ({ children }) => {
  return <DivWrapper>{children}</DivWrapper>;
};

export default ProductContent;
