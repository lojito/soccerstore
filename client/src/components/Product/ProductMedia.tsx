import { type FC } from "react";
import { FaHeart } from "react-icons/fa";
import { MdOutlineShoppingCart } from "react-icons/md";
import { useDispatch, useSelector } from "react-redux";
import { styled } from "styled-components";
import { addToCart, removeFromCart } from "../../actions/cart";
import { addFavorite, removeFavorite } from "../../actions/favorite";
import useCreateFavorite from "../../hooks/favorite/useCreateFavorite";
import useDeleteFavorite from "../../hooks/favorite/useDeleteFavorite";
import { RootReducerType } from "../../reducers";

const DivWrapper = styled.div`
  position: relative;
  width: 225px;
  margin: 0 auto;

  & svg.favorite {
    position: absolute;
    right: 15px;
    top: 10px;
    color: red;

    @media screen and (max-width: 1160px) {
      right: 10%;
    }
  }

  & svg.favorite.add {
    color: green;
  }

  & svg.cart {
    position: absolute;
    left: 15px;
    top: 10px;
    color: red;

    @media screen and (max-width: 1160px) {
      left: 10%;
    }
  }

  & svg.cart.added {
    color: green;
  }

  & svg:hover {
    cursor: pointer;
  }

  & image {
    height: 100%;
  }
`;

type Props = {
  id: number;
  image: string;
  isLoggedIn: boolean;
  token: string | null;
};

const ProductMedia: FC<Props> = ({ id, image, isLoggedIn, token }) => {
  const dispatch = useDispatch();

  const productsInFavoritesList = useSelector(
    (state: RootReducerType) => state.favorites
  );
  const isProductInFavoritesList = productsInFavoritesList.includes(id);

  const iconClass = isProductInFavoritesList ? "favorite add" : "favorite";
  const title = isProductInFavoritesList
    ? "Elimina este producto de tu lista de favoritos."
    : "Adiciona este producto a tu lista de favoritos.";

  const { handleCreateFavorite } = useCreateFavorite(token!);
  const { handleDeleteFavorite } = useDeleteFavorite(token!);

  const handleFavorite = async () => {
    if (isProductInFavoritesList) {
      await handleDeleteFavorite(id);
      dispatch(removeFavorite(id));
    } else {
      await handleCreateFavorite(id);
      dispatch(addFavorite(id));
    }
  };

  const productsInCart = useSelector((state: RootReducerType) => state.cart);
  const isProductInCart = productsInCart.includes(id);

  const cartClass = isProductInCart ? "cart added" : "cart";
  const cartTitle = isProductInCart
    ? "Elimina este producto del carrito."
    : "Adiciona este producto al carrito.";

  const handleShoppingCart = async () => {
    if (isProductInCart) {
      dispatch(removeFromCart(id));
    } else {
      dispatch(addToCart(id));
    }
  };

  return (
    <DivWrapper>
      {isLoggedIn && (
        <>
          <FaHeart
            className={iconClass}
            onClick={handleFavorite}
            title={title}
            aria-label="favorite-icon"
            role="img"
          />
          <MdOutlineShoppingCart
            onClick={handleShoppingCart}
            className={cartClass}
            title={cartTitle}
            aria-label="cart-icon"
            role="img"
          />
        </>
      )}
      <img src={image} />
    </DivWrapper>
  );
};

export default ProductMedia;
