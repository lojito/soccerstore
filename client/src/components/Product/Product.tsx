import { type FC } from "react";
import { styled } from "styled-components";
import ProductContent from "./ProductContent";
import ProductMedia from "./ProductMedia";

const DivWrapper = styled.div`
  width: 20%;
  display: inline-block;
  margin-bottom: 20px;
  text-align: center;
  vertical-align: top;

  @media screen and (max-width: 1160px) {
    width: 25%;
  }

  @media screen and (max-width: 970px) {
    width: 33%;
  }

  @media screen and (max-width: 740px) {
    width: 50%;
  }

  @media screen and (max-width: 490px) {
    width: 100%;
  }
`;

export type ProductType = {
  id: number;
  category: string;
  name: string;
  price: string;
  image: string;
};

type Props = {
  product: ProductType;
  isLoggedIn: boolean;
  token: string | null;
};

const Product: FC<Props> = ({
  product: { id, name, price, image, category },
  isLoggedIn,
  token,
}) => {
  return (
    <DivWrapper>
      <ProductMedia
        id={id}
        image={`images/${category}/${image}.jpg`}
        isLoggedIn={isLoggedIn}
        token={token}
      />
      <ProductContent>{name}</ProductContent>
      <ProductContent>{price} MXN</ProductContent>
    </DivWrapper>
  );
};

export default Product;
