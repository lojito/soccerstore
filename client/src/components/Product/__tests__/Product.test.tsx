import { render } from "@testing-library/react";
import Root from "../../../Root";
import Product from "../Product";
import { Props as ProductContentProps } from "../ProductContent";
import { Props as ProductMediaProps } from "../ProductMedia";

vi.mock("../ProductContent", () => ({
  default: (props: ProductContentProps) => {
    return <div>{props.children}</div>;
  },
}));

vi.mock("../ProductMedia", () => ({
  default: (props: ProductMediaProps) => {
    return <div>{props.image}</div>;
  },
}));

describe("ProductContent", () => {
  it("should render correctly", () => {
    const product = {
      id: 1,
      category: "balones",
      name: "Balon Blanco y Negro",
      price: "29.99",
      image: "1",
    };

    const { asFragment } = render(
      <Root>
        <Product isLoggedIn={false} token="" product={product} />
      </Root>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
