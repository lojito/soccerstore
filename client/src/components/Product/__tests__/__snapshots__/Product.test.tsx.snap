// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductContent > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-beySPh ODEZD"
  >
    <div>
      images/balones/1.jpg
    </div>
    <div>
      Balon Blanco y Negro
    </div>
    <div>
      29.99 MXN
    </div>
  </div>
</DocumentFragment>
`;
